New cmdlets for working with subscriptions

markwragg opened the pull request:
**Changes proposed in this pull request:**
 - Add a cmdlet New-RsSubscription for creating new subscriptions via PowerShell
 - Add a cmdlet New-RsScheduleXML for creating the XML string definition of a schedule needed to work with New-RsSubscription -Schedule parameter
 - Add a cmdlet Export-RsSubscription for exporting subscriptions to XML via Export-CliXML. This function isn't strictly needed but its there to compliment Import-RsSubscription and make explicit how Subscriptions need to be stored (e.g in XML as generated by Export-CliXML).
- Add a cmdlet Import-RsSubscription for importing subscriptions that have exported to disk as XML and recreating their .ExtensionSettings property so that it is a valid object for use with the existing Set-RsSubscription cmdlet.

4th October 2017 Changes:
- Modified Set-RsSubscription so that it recreates the DeliverySettings property for Subscription objects that have been deserialized (e.g if you have Export/Imported them via Import-CliXML).

5th October 2017 Changes:
- Modified Get-RsSubscription to support the 2005 SOAP webservice API via a parameter so that you can retrieve subscriptions from old versions of SSRS.
- Modified New-RsWebServiceProxy and Connect-RsReportServer to support the -ApiVersion switch and settings so that the old webservice URL can be queried when needed. 
- Have added Pester tests for each new function.
- Have amended the Readme.md regarding new functions, as well as added `Get-RsSubscription` and `Set-RsSubscription` to Readme.md as these were missing.

10th October 2017 Changes:
- This commit addresses all of the comments made on the PR by parthsha.
- This commit reintroduces the Import and Export cmdlets as Import-RsSubscriptionXml and Export-RsSubscriptionXml. Decided these were needed to enforce -Depth 3 on export and to move the recreation of the deserialized objects out of the Set-RsSubscription cmdlet so as to not convolute it.
- Reintroduces pester tests for Import- and Export-RsSubscription
- Fixes a bug in Get-RsSubscription where $item was being set in a If statement instead of tested (line 81)
- Changes Get-RsSubscription so that it can retrieve data driven subscriptions.
- Changes Set-RsSubscription so that it can create data driven subscriptions.

**How to test this code:**
- Run the relevant Pester tests
 - Manually create a new subscription via `New-RsSubscription -Path /some/path -Description 'This is a test' -Destination FileShare -DestinationPath '\\some\path' -Filename 'Somefile' -Schedule (New-RsScheduleXML) -RenderFormat PDF`
 - Export subscriptions via: `Get-RsSubscription | Export-RsSubscriptions .\sub.xml`
 - Import subscriptions and add to SRSS via `Import-RsSubscription .\sub.xml | Set-RsSubscription -Path /some/path`

**Has been tested on (remove any that don't apply):**
 - Powershell 5.1
 - Windows Server 2016
 - SQL Server 2016
- SRSS Version 13.0.4435.0

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/ReportingServicesTools?pullRequest=79) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/ReportingServicesTools?pullRequest=79) before we can accept your contribution.<br/><br/>:x: markwragg [sign now](https://cla.opensource.microsoft.com/Microsoft/ReportingServicesTools?pullRequest=79)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/ReportingServicesTools?pullRequest=79) it.</sub>

markwragg commented:
I discussed these new functions with @ClaudioESSilva on Twitter and he suggested that the Import- Export- functions could just be worked in to the existing functions. Thinking about if I put the fix that the Import- cmdlet performs in to the Set-RsSubscription cmdlet then the Import- and Export- cmdlets can just be replaced with Import/Export-CliXML. I'm going to have a go at doing that.

markwragg commented:
I have now done as noted in the previous comment, I've amended the Set-RsSubscription cmdlet so that it recreates the `.DeliverySettings` property for a Deserialized object and removed the Import/Export-RsSubscription cmdlets that I added (as well as their Pester tests and references to them in Readme.md). I've also added a test to Set-RsSubscription.Tests.ps1 to show that it can create a subscription that has been saved and loaded from disk via Export/Import-CliXML.

As a result you should now be able to:
```
Get-RsSubscription -Path /some/path | Export-CliXML .\somefile.xml
Import-CliXML .\somefile.xml | Set-RsSubscription -Path /some/path
```
Successfully.

This PR of course still contains the the new `New-RsSubscription` cmdlet which I think will be useful to anyone looking to create subscriptions from scratch directly with PowerShell.

parthsha commented on ReportingServicesTools/Functions/CatalogItems/Get-RsSubscription.ps1:
```diff
@@ -52,7 +57,11 @@ function Get-RsSubscription
+
+        [ValidateSet('2005','2010')]
+        [string]
+        $APIVersion = '2010'
```
Please rename this to ApiVersion.

parthsha replied:
Could you also please move ApiVersion after ReportServerUri?

parthsha replied:
Any reason why 2006 is missing from this list?

markwragg replied:
I wasn't sure whether it would work against the 2006 API, as it seemed to be a completely different set of methods (I read that the 2006 API is geared around working with Sharepoint). I'll add it though and test.

parthsha commented on ReportingServicesTools/Functions/Common/ConnectionObjectRequests.ps1:
```diff
@@ -35,7 +36,7 @@
     }
     
-    $goodKeys = @("ReportServerUri", "Credential")
+    $goodKeys = @("ReportServerUri", "Credential","APIVersion")
```
Please rename this to ApiVersion.

markwragg replied:
Done, will commit the change shortly.

parthsha commented on ReportingServicesTools/Functions/CatalogItems/Get-RsSubscription.ps1:
```diff
@@ -26,6 +26,11 @@ function Get-RsSubscription
             Use "New-RsWebServiceProxy" to generate a proxy object for reuse.
             Useful when repeatedly having to connect to multiple different Report Server.
         
+        .PARAMETER APIVersion
```
Please rename this to ApiVersion.

parthsha replied:
Could also please move this after .PARAMETER ReportServerUri?

markwragg replied:
Done, will commit the change shortly.

parthsha commented on ReportingServicesTools/Functions/Utilities/Connect-RsReportServer.ps1:
```diff
@@ -97,7 +106,7 @@ function Connect-RsReportServer
         try
         {
-            $proxy = New-RsWebServiceProxy -ReportServerUri ([Microsoft.ReportingServicesTools.ConnectionHost]::ReportServerUri) -Credential ([Microsoft.ReportingServicesTools.ConnectionHost]::Credential) -ErrorAction Stop
+            $proxy = New-RsWebServiceProxy -ReportServerUri ([Microsoft.ReportingServicesTools.ConnectionHost]::ReportServerUri) -Credential ([Microsoft.ReportingServicesTools.ConnectionHost]::Credential) -ApiVersion $APIVersion -ErrorAction Stop
```
Please rename APIVersion to ApiVersion.

markwragg replied:
Done, will commit the change shortly.

parthsha commented on ReportingServicesTools/Functions/Utilities/New-RsWebServiceProxy.ps1:
```diff
@@ -23,6 +23,11 @@ function New-RsWebServiceProxy
             Specify the Credential to use when connecting to your SQL Server Reporting Services Instance.
             By Default, this parameter uses the credentials (if any) specified by the Connect-RsReportingServer function.
         
+        .PARAMETER APIVersion
```
Please rename this to ApiVersion.

markwragg replied:
Done, will commit the change shortly.

parthsha commented on ReportingServicesTools/Functions/Utilities/New-RsWebServiceProxy.ps1:
```diff
@@ -51,7 +56,11 @@ function New-RsWebServiceProxy
+
+        [ValidateSet('2005','2006','2010')]
+        [string]
+        $APIVersion = '2010'
```
Please rename this to ApiVersion.

markwragg replied:
Done, will commit the change shortly.

parthsha commented on ReportingServicesTools/Functions/Utilities/New-RsWebServiceProxy.ps1:
```diff
@@ -84,7 +93,7 @@ function New-RsWebServiceProxy
     }
     $reportServerUriObject = New-Object System.Uri($ReportServerUri)
-    $soapEndpointUriObject = New-Object System.Uri($reportServerUriObject, 'ReportService2010.asmx')
+    $soapEndpointUriObject = New-Object System.Uri($reportServerUriObject, "ReportService$APIVersion.asmx")
```
Please rename APIVersion to ApiVersion.

markwragg replied:
Done, will commit the change shortly.

parthsha commented on ReportingServicesTools/ReportingServicesTools.psd1:
```diff
@@ -98,7 +98,9 @@
         'Write-RsFolderContent',
-        'Write-RsRestCatalogItem'
+        'Write-RsRestCatalogItem',
+        'New-RsSubscription',
```
Could you please move the New-RsSubscription and New-RsScheduleXml after New-RsRestSession and before New-RsWebServiceProxy?

markwragg replied:
Done.

parthsha commented on README.md:
```diff
@@ -43,6 +44,8 @@ The following is a list of commands which are available for you to use once you
 |New-RsFolder|This command creates a new folder in the specified path.|
 |New-RsRestSession|This command creates a session object to be specified for all subsequent calls to the REST Endpoint.|
 |New-RsWebServiceProxy|This command creates a new Web Service Proxy which is used to interact with the SOAP Endpoint.|
+|New-RsSubscription|This command adds a new subscription to an existing report.|
```
Please move this before New-RsWebServiceProxy.

markwragg replied:
Done

parthsha commented on README.md:
```diff
@@ -43,6 +44,8 @@ The following is a list of commands which are available for you to use once you
 |New-RsRestSession|This command creates a session object to be specified for all subsequent calls to the REST Endpoint.|
 |New-RsWebServiceProxy|This command creates a new Web Service Proxy which is used to interact with the SOAP Endpoint.|
+|New-RsSubscription|This command adds a new subscription to an existing report.|
+|New-RsScheduleXML|This command creates an XML string definition of a subscription schedule. For use with the -Schedule parameter or New-RsSubscription.|
```
Please move this before New-RsWebServiceProxy.

markwragg replied:
Done

parthsha commented on ReportingServicesTools/Functions/Utilities/Connect-RsReportServer.ps1:
```diff
@@ -32,6 +32,11 @@ function Connect-RsReportServer
         .PARAMETER ReportPortalUri
             The Uri to connect to for accessing the REST Endpoint. This exists in SQL Server Reporting Services 2016 and later.
         
+        .PARAMETER APIVersion
```
It is a bit strange that this parameter is added to this cmdlet because it is not really doing anything. Every single parameter to this cmdlet modifies Microsoft.ReportingServicesTools.ConnectionHost (which is defined in <root>\ReportingServicesTools\Libraries\library.ps1). I recommend either removing this parameter altogether or adding ApiVersion in ConnectionHost. My only concern with adding something like ApiVersion in ConnectionHost is there are 2 ways to communicate with Reporting Services starting from 2016: SoapEndpoint found at ReportServerUri and RestEndpoint found at PortalUri (which is open to developers starting from SSRS 2017). If you do add ApiVersion to ConnectionHost, I highly recommend you to call it SoapEndpointApiVersion, so that people don't confuse it with the ApiVersion of RestEndpoint.

markwragg replied:
I think its needed as a parameter to pass through to New-RsWebServiceProxy (although maybe this could be New-RsWebServiceProxyHelper instead?). Anyway i've gone with the solution of renaming it to $SoapEndpointApiVersion as suggested.

parthsha commented on ReportingServicesTools/Functions/CatalogItems/New-RsScheduleXML.ps1:
```diff
@@ -0,0 +1,198 @@
+﻿# Copyright (c) 2016 Microsoft Corporation. All Rights Reserved.
+# Licensed under the MIT License (MIT)
+
+Function New-RsScheduleXML {
```
Please rename this to New-RsScheduleXml

markwragg replied:
Done, will commit the change shortly.

parthsha commented on ReportingServicesTools/Functions/CatalogItems/New-RsScheduleXML.ps1:
```diff
@@ -0,0 +1,198 @@
+        $End
+
+    )
+    Process {
```
Please format using the guidelines we have specified in Readme. Most of the code is easy to read except for the statements where there are multiple | combined with complex logic.

markwragg replied:
Overhauled in next commit to meet style guide and is now much more readable.

parthsha commented on ReportingServicesTools/Functions/CatalogItems/New-RsSubscription.ps1:
```diff
@@ -0,0 +1,216 @@
+        .PARAMETER Path
+            Specify the path to the destination report.
+        
+        .PARAMETER ReportServerUri
```
For consistency sake, I recommend you to add ApiVersion parameter to this cmdlet also.

markwragg replied:
Not sure it yet supports the older APIs for creating subscriptions so am going to leave this for now.

parthsha commented on ReportingServicesTools/Functions/CatalogItems/Set-RsSubscription.ps1:
```diff
@@ -90,6 +90,7 @@ function Set-RsSubscription
     Begin
```
For consistency sake, I would recommend you to add ApiVersion to this cmdlet also.

markwragg replied:
Not sure it yet supports the older APIs for creating subscriptions so am going to leave this for now.

markwragg commented on ReportingServicesTools/Functions/CatalogItems/Get-RsSubscription.ps1:
```diff
@@ -67,7 +76,16 @@ function Get-RsSubscription
-                $subscriptions = $Proxy.ListSubscriptions($Item)
+                if ($Proxy.Url -match 'ReportService2005.asmx') 
+                {
+                    If ($item = '/') { $item = $null }
```
Used = instead of -eq. Will be fixed in next commit.
